test/timed_asset_lock_contract_test.py

```diff
@@ -79,112 +89,300 @@ def app_id(test_config, asset_id, end_time, wallet_1):
+    private_key = wallet['privateKey']
+    if adversary_wallet:
+        # try to opt in as advesary
+        txn, txn_id = opt_in_app_signed_txn(private_key,
+                                            public_key,
+                                            params,
+                                            app_id,
+                                            foreign_assets=[asset_id])
+        with pytest.raises(AlgodHTTPError):
+            client.send_transactions([txn])
+            wait_for_txn_confirmation(client, txn_id, 5)
+
+        # attempt to call application (on_setup function)
+        txn, txn_id = noop_app_signed_txn(private_key, public_key, params, app_id, [asset_id])
+        with pytest.raises(AlgodHTTPError):
+            client.send_transactions([txn])
+            wait_for_txn_confirmation(client, txn_id, 5)
+
+
+    # attempt to delete application
+    txn, txn_id = delete_app_signed_txn(private_key, public_key, params, app_id, [asset_id])
+    with pytest.raises(AlgodHTTPError):
+        client.send_transactions([txn])
+        wait_for_txn_confirmation(client, txn_id, 5)
+
+    # attempt to update the program
+    new_approval_program = compileTeal(Approve(), Mode.Application, version=5)
+    new_clear_program = compileTeal(Approve(), Mode.Application, version=5)
+
+    new_approval_program = compile_program(client, new_approval_program)
+    new_clear_program = compile_program(client, new_clear_program)
+    txn, txn_id = update_app_signed_txn(private_key,
+                                        public_key,
+                                        params,
+                                        app_id,
+                                        new_approval_program,
+                                        new_clear_program)
+    with pytest.raises(AlgodHTTPError):
+        client.send_transactions([txn])
+        wait_for_txn_confirmation(client, txn_id, 5)
+
+    # attempt to close out application
+    txn, txn_id = close_out_app_signed_txn(private_key,
+                                           public_key,
+                                           params,
+                                           app_id,
+                                           [asset_id])
+    with pytest.raises(AlgodHTTPError):
+        client.send_transactions([txn])
+        wait_for_txn_confirmation(client, txn_id, 5)
+
+    # attempt to clear application state
+    txn, txn_id = clear_state_out_app_signed_txn(private_key,
+                                                 public_key,
+                                                 params,
+                                                 app_id,
+                                                 [asset_id])
+    if adversary_wallet or fail_clear:
+        with pytest.raises(AlgodHTTPError):
+            client.send_transactions([txn])
+            wait_for_txn_confirmation(client, txn_id, 5)
+
+
+def cash_out(client, public_key, private_key, app_id, asset_ids):

```

Comment:
should these functions be in testing_utils.py?

cash_out, opt_out, opt_in, set_up